perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu.java

```diff
@@ -0,0 +1,142 @@
+ * <li>no duplicates among the list values</li>
+ * <li>no duplicates among all values of this attribute and all values of
+ * attribute urn:perun:member:attribute-def:def:o365EmailAddresses:mu</li>
+ * </ul>
+ *
+ * @author Michal Stava &lt;stavamichal@gmail.com&gt;
+ */
+public class urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu extends GroupAttributesModuleAbstract implements GroupAttributesModuleImplApi {
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu.class);
+	static final String ADNAME_ATTRIBUTE = AttributesManager.NS_GROUP_ATTR_DEF + ":adName:o365mu";
+
+	@Override
+	public void checkAttributeSyntax(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongAttributeValueException {
+		log.trace("checkAttributeSyntax(group={},attribute={})", group, attribute);
+		List<String> emails = attribute.valueAsList();
+
+		if (emails == null) return;
+
+		//check syntax of all values
+		for (String email : emails) {
+			Matcher emailMatcher = emailPattern.matcher(email);
+			if (!emailMatcher.matches())
+				throw new WrongAttributeValueException(attribute, group, "Email " + email + " is not in correct form.");
+		}
+
+		//check for duplicities
+		if (hasDuplicate(emails)) {
+			throw new WrongAttributeValueException(attribute, group, "duplicate values");
+		}
+	}
+
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		List<String> emails = attribute.valueAsList();
+
+		if (emails == null) emails = new ArrayList<>();
+		//at least one value if adName is set
+		AttributesManagerBl am = sess.getPerunBl().getAttributesManagerBl();
+		try {
+			if (emails.isEmpty()) {
+				Attribute adName = am.getAttribute(sess, group, ADNAME_ATTRIBUTE);
+				if (adName.getValue() != null) {
+					throw new WrongReferenceAttributeValueException(attribute, adName, group, null, group, null, "at least one email must be defined");
+				}
+
+				return;
+			}
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e.getMessage(), e);
+		}
+
+		//check for duplicities among all members and groups
+		AttributesManagerBl attributesManagerBl = sess.getPerunBl().getAttributesManagerBl();
+		Set<Integer> setOfGroupsDefinedByAttribute = attributesManagerBl.getPerunBeanIdsForUniqueAttributeValue(sess, attribute).stream().map(Pair::getLeft).collect(Collectors.toSet());
+		setOfGroupsDefinedByAttribute.remove(group.getId());
+
+		if (!setOfGroupsDefinedByAttribute.isEmpty()) {
+			throw new WrongReferenceAttributeValueException(attribute, attribute, group, null, "some of the email addresses are already assigned to the following group pairs: " + setOfGroupsDefinedByAttribute);
+		}
+		Attribute memberO365EmailAddresses = new Attribute(new urn_perun_member_attribute_def_def_o365EmailAddresses_mu().getAttributeDefinition());
```

Comment:
I'm not sure we rely on attribute definition from the module except the tests, since it might differ from the real perun instance state. Also subsequent call of ``getPerunBeanIdsForUniqueAttributeValue()`` requires attribute definition to have real id, which can't be present in this construct as is set to 0.
I suggest to get attribute definition by name (urn).

Reply:
Changed.